libbeat/outputs/shipper/shipper.go

```diff
@@ -156,28 +178,41 @@ func (c *shipper) Publish(ctx context.Context, batch publisher.Batch) error {
 		})
 	}
 
+	st.Dropped(dropped)
 	c.log.Debugf("%d events converted to protobuf, %d dropped", len(grpcEvents), dropped)
 
 	ctx, cancel := context.WithTimeout(ctx, c.timeout)
 	defer cancel()
-	_, err := c.client.PublishEvents(ctx, &sc.PublishRequest{
+	resp, err := c.client.PublishEvents(ctx, &sc.PublishRequest{
 		Events: grpcEvents,
 	})
 
-	if err != nil {
-		if status.Code(err) == codes.ResourceExhausted {
-			c.log.Warn("shipper's queue is full, more events cannot be accepted")
-			batch.Cancelled()
-		} else {
-			batch.Retry()
-		}
-		return fmt.Errorf("failed to publish the batch to the shipper: %w", err)
-	}
+	switch {
 
-	batch.ACK()
+	case resp == nil:
+		// if the response is nil, it means none of the events made it through due to a severe failure
+		batch.Cancelled()
+		st.Cancelled(len(events))
+		c.log.Errorf("failed to publish the batch to the shipper, none of the %d events were accepted: %w", len(grpcEvents), err)
+		return nil
```

Comment:
Right, you likely need to return a successful response and then infer that the success was partial from the content of the response message.

Reply:
> Reading the implementation of findRetries I'm not sure I like that we have to do that for every successful message. Perhaps we should change the protocol and optimize for the most common case where all events were published.

After changes in https://github.com/elastic/beats/pull/31558/commits/40e4d3fb5f9c0117f1482148639464b927f8539b it's not that demanding anymore, just comparing the slice sizes basically and then creating a slice on the existed memory.

> We could make it so that if the entire batch isn't accepted into the queue, it is is all rejected and an error with codes.ResourceExhausted is returned. The simplest approach is to just retry the whole batch.

This would not break the current implementation here, both cases are supported. And this logic is already present.

> If we wanted to optimize from there we could have the client here retry with a smaller batch size

I don't think it should be a responsibility of the client (output implementation) the event pipeline should decide about restructuring the batches, we already have this need for Elasticsearch output and its payload size limitation. So, I hope our new pipeline will have this feature.